doc/src/tutorial/codegen.rst

```diff
@@ -0,0 +1,516 @@
+===============
+Code Generation
+===============
+Several submodules in Sympy are provided to generate directly compilable 
+code from Sympy expressions. Furthermore, there are submodules provided 
+that help use other numeric systems as backend for construction and 
+manipulation of mathematical expressions to speed up the process.
+
+We will start with a brief introduction, and inspect the building 
+blocks of the code generation and then the code generation itself.
+
+Introduction
+------------
+
+There are three levels::
+
+    autowrap
+       |
+    codegen
+       |
+    code printers
+
+Autowrap uses codegen, and codegen uses the code printers. Autowrap does 
+everything: it lets you go from SymPy expression to numerical 
+function in the same Python process in one step. codegen is actual 
+code generation, i.e., to compile and use later, or to include in some larger 
+project.
+
+The code printers translate the SymPy objects into actual code, 
+like abs(x) -> fabs(x) (for C).
+
+The code printers don't print optimal code in many cases. 
+An example of this is powers in C. x**2 prints as pow(x, 2) instead of x*x. 
+Other optimizations (like mathematical simplifications) should happen 
+before the code printers.
+
+cse is not applied anywhere in this chain. It ideally happens at the 
+codegen level, or somewhere above it.
+
+We will iterate through the levels, bottom up.
+
+Code printers (sympy.printing)
+------------------------------
+This is where the meat of code generation is, the translation of SymPy
+expressions to specific code.Supported languages are C (``ccode``), Fortran 95 (``fcode``), 
+Javascript (``jscode``), Mathematica (``mathematica_code``), Octave/Matlab (``octave_code``), 
+Python (print_python, which is actually more like a lightweight version 
+of codegen for Python, and ``lambdarepr``, which supports many libraries 
+(like NumPy), and theano (``theano_code``).The code printers 
+are special cases of the other prints in SymPy (str printer, pretty printer, etc.).
+
+An important distinction is that the code printer has to deal with 
+assignments (using the ``sympy.printing.codeprinter.Assignment`` object).This serves
+as building blocks for the code printers and hence the ``codegen`` module.
+An example that shows the use of ``Assignment``::
+
+    >>> from sympy import symbols, MatrixSymbol, Matrix
+    >>> from sympy.printing.codeprinter import Assignment
+    >>> x, y, z = symbols('x y z')
+    >>> mat = Matrix([x, y, z]).T
+    >>> known_mat = MatrixSymbol('K', 1, 3)
+    >>> Assignment(known_mat, mat)  # doctest: +SKIP
```

Comment:
Pleaes remove all of the skip statements and send a new pull request so we can see what is failing on Travis. In general, the solution for failures on Travis is never going to be to skip the test. That defeats the purpose of testing.